### PR TITLE
Detach objects in EF to reduce memory usage

### PR DESCRIPTION
**Pull request by mocsharp**
### Description
* Reduce memory usage by explicitly detaching objects upon adding or updating with status of complete/failure.
* Update Dockerfile to include dotnet debugging tools.
* Removes files from disk after job submission reaches maximum retries.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] All tests passed locally by running `./src/run-tests-in-docker.sh`.
- [ ] [Documentation comments](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments) included/updated.
- [ ] User guide updated.
- [ ] NGC publishing metadata updated.
- [ ] I have updated the changelog

### CODE REVIEW

File: src/API/IJobsRepository.cs [hunk 1]
```diff
@@ -39,6 +39,14 @@ public interface IJobRepository
         /// <param name="job"><see cref="Nvidia.Clara.DicomAdapter.API.InferenceJob" /></param>
         Task Add(InferenceJob job);
 
+        /// <summary>
+        /// Queues a new job for submission and disable EF change tracking.
+        /// <c>New</c> makes a copy of the instances to a temporary location that is to be
+        /// uploaded by the `Nvidia.Clara.DicomAdapter.Server.Services.Jobs.JobSubmissionService`.
+        /// </summary>
+        /// <param name="job"><see cref="Nvidia.Clara.DicomAdapter.API.InferenceJob" /></param>
+        Task AddWithoutTracking(InferenceJob job);
```
Comment (MMelQin) [inline src/API/IJobsRepository.cs:48]: Should the Add method simply be updated to include tracking flag param? The implementation class below ends up calling the AddJob method with the flag param, wonder why having two separate method at the I/F level.

File: src/Server/Processors/AeTitleJobProcessor.cs [hunk 2]
```diff
@@ -208,9 +209,15 @@ private void OnTimedEvent(Object source, System.Timers.ElapsedEventArgs e)
...
-                            _instances.Remove(key);
-                            _logger.Log(LogLevel.Information, $"Timeout elapsed waiting for {_grouping} {key}");
+                            if (_jobs.TryAdd(_instances[key]))
+                            {
+                                _instances.Remove(key);
+                                _logger.Log(LogLevel.Information, $"Timeout elapsed waiting for {_grouping} {key}");
+                            }
+                            else
+                            {
+                                _logger.Log(LogLevel.Error, $"Error while queuing new instance collection with key={key}, grouping={_grouping}.");
```
Comment (MMelQin) [inline src/Server/Processors/AeTitleJobProcessor.cs:219]: Could the message a little more specific, like "Failed to enqueue ...."?

File: src/Server/Repositories/DicomAdapterRepository.cs [hunk 3]
```diff
@@ -110,5 +115,11 @@ public T FirstOrDefault(Func<T, bool> func)
 
             return _dicomAdapterContext.Set<T>().FirstOrDefault(func);
         }
+
+        public void Detach(T item)
+        {
+            Guard.Against.Null(item, nameof(item));
+            _dicomAdapterContext.Entry(item).State = EntityState.Detached;
```
Comment (MMelQin) [inline src/Server/Repositories/DicomAdapterRepository.cs:122]: What if item does not exist in Entry?

Comment (mocsharp) [reply src/Server/Repositories/DicomAdapterRepository.cs:122]: Looking at the source code to `Entry`, it looks like it'll create a new item, which is fine in our use case.

File: src/API/IJobsRepository.cs (hunk 1 above)
Comment (mocsharp) [reply src/API/IJobsRepository.cs]: Updated

File: src/Server/Processors/AeTitleJobProcessor.cs (hunk 2 above)
Comment (mocsharp) [reply src/Server/Processors/AeTitleJobProcessor.cs]: Updated

**Review by whoisj — DISMISSED** [review]
Overall it looks fine to me. I'm still not clear on the overall threading model here, but I've been confused by it for a long time now 😏

File: Dockerfile [hunk 4]
```diff
@@ -18,6 +18,11 @@ FROM mcr.microsoft.com/dotnet/core/sdk:3.1-bionic as build
 ARG Version=0.0.0
 ARG FileVersion=0.0.0.0
 
+# Install the tools
+RUN dotnet tool install --tool-path /tools dotnet-trace
```
[the remark this answers is not in the document]
Comment (mocsharp) [reply Dockerfile:22]: yes, easier to debug when things happen.

**Review by whoisj — APPROVED** [review]

**Pull request merged**